src/components/tl/nccl/allgatherv/allgatherv.c

```diff
@@ -93,6 +93,244 @@ ucc_status_t ucc_tl_nccl_allgatherv_p2p_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_nccl_task_t *task = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+    ucc_coll_args_t *args = &TASK_ARGS(task);
+    ucc_tl_nccl_team_t *team = TASK_TEAM(task);
+    size_t sdt_size = ucc_dt_size(args->src.info.datatype);
+    size_t rdt_size = ucc_dt_size(args->dst.info_v.datatype);
+
+    /* initialize variables, so we don't accidentally use garbage
+     * values */
+    win->cpu_sbuf = NULL;
+    win->staged_sbuf = NULL;
+    win->sbuf_len = 0;
+    win->first_peer_rank = -1;
+    win->first_peer_cpu_rbuf = NULL;
+    win->first_peer_len = 0;
+    win->last_peer_rank = -1;
+    win->last_peer_len = 0;
+
+    uintptr_t window_start = round * UCC_TL_NCCL_SCRATCH_BUF_SIZE;
+    uintptr_t window_end = window_start + UCC_TL_NCCL_SCRATCH_BUF_SIZE;
+
+
+    /* sbuf setup */
+    uintptr_t sbuf_start = 0;
+    for (int peer = 0; peer < UCC_TL_TEAM_RANK(team); peer++) {
+        sbuf_start += ucc_coll_args_get_count(args, args->dst.info_v.counts, peer) * rdt_size;
+    }
+    uintptr_t sbuf_end = sbuf_start + args->src.info.count * sdt_size;
+
+    if (sbuf_end > window_start && sbuf_start < window_end) {
+        uintptr_t sbuf_offset = 0;
+        if (sbuf_start < window_start) {
+            sbuf_offset = window_start - sbuf_start;
+        }
+
+        win->cpu_sbuf = PTR_OFFSET(args->src.info.buffer, sbuf_offset);
+        if (sbuf_start <= window_start) {
+            win->staged_sbuf = task->cpu_coll_scratch_buf;
+        } else {
+            win->staged_sbuf = PTR_OFFSET(task->cpu_coll_scratch_buf, sbuf_start - window_start);
+        }
+        win->sbuf_len = MIN(sbuf_end, window_end) - MAX(sbuf_start, window_start);
+    }
+
+
+    /* rbuf setup */
+    uintptr_t offset = 0;
+    int first_peer = 1;
+    for (int peer = 0; peer < UCC_TL_TEAM_SIZE(team); peer++) {
+        uintptr_t recv_size = ucc_coll_args_get_count(args, args->dst.info_v.counts, peer) * rdt_size;
+
+        if (recv_size == 0) {
+            continue;
+        } else if (offset + recv_size < window_start) {
+            offset += recv_size;
+            continue;
+        } else if (offset >= window_end) {
+            break;
+        }
+
+        recv_size = MIN(offset + recv_size, window_end) - MAX(offset, window_start);
+
+        if (first_peer) {
+            win->first_peer_rank = peer;
+            uintptr_t displ = ucc_coll_args_get_displacement(args, args->dst.info_v.displacements, peer);
+            win->first_peer_cpu_rbuf = PTR_OFFSET(args->dst.info_v.buffer, displ * rdt_size);
+            win->first_peer_len = recv_size;
+
+            first_peer = 0;
+        }
+
+        win->last_peer_rank = peer;
+        win->last_peer_len = recv_size;
+
+        offset += recv_size;
+    }
+}
+
+static void CUDART_CB cpu_allgatherv_copy_in(void *data)
+{
+    ucc_coll_task_t *coll_task = (ucc_coll_task_t *) data;
+    ucc_tl_nccl_task_t *task = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+
```

Comment:
window_bounds_t win; goes together with variables decl and empty line goes after.